probe-rs/src/flashing/builder.rs

```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
     ///
-    /// Programming does not start until the `program` method is called.
-    pub(super) fn add_data(&mut self, address: u32, data: &[u8]) -> Result<(), FlashError> {
-        // Add the operation to the sorted data list.
-        match self
-            .data_blocks
-            .binary_search_by_key(&address, |v| v.address)
-        {
-            // If it already is present in the list, this indicates a bug in the flashing code.
-            Ok(_) => panic!(
-                        "Error preparing data to flash. Address {0:#010x} was already written earlier. This is a bug, please report it.",
-                        address),
-            // Add it to the list if it is not present yet.
-            Err(position) => {
-                // If we have a prior block (prevent u32 underflow), check if its range intersects
-                // the range of the block we are trying to insert. If so, return an error.
-                if position > 0 {
-                    if let Some(block) = self.data_blocks.get(position - 1) {
-                        let range = block.address..block.address + block.data.len() as u32;
-
-                        assert!(
-                            !range.intersects_range(&(address..address + data.len() as u32)),
-                            "Overlap in data, address {0:#010x} was already written earlier. This is a bug, please report it.",
-                            address
-                        );
-                    }
-                }
+    /// The chunk can cross flash boundaries as long as one flash region connects to another flash region.
+    pub fn add_data(&mut self, address: u32, data: &[u8]) -> Result<(), FlashError> {
+        // Ignore zero-length stuff
+        if data.is_empty() {
+            return Ok(());
+        }
 
-                // If we have a block after the one we are trying to insert,
-                // check if its range intersects the range of the block we are trying to insert.
-                // If so, return an error.
-                // We don't add 1 to the position here, because we have not insert an element yet.
-                // So the ones on the right are not shifted yet!
-                if let Some(block) = self.data_blocks.get(position) {
-                    let range = block.address..block.address + block.data.len() as u32;
-
-                    assert!(
-                        !range.intersects_range(&(address..address + data.len() as u32)),
-                        "Error preparing data to flash. Address {0:#010x} is not a valid address in the flash area. This is a bug, please report it.",
-                        address
-                    );
-                }
+        // Check the new data doesn't overlap to the right
+        if let Some((&next_addr, _)) = self.data.range(address..).next() {
+            if address + (data.len() as u32) > next_addr {
+                panic!("data overlaps to the right");
```

Comment:
Since you made this function public I am not okay with the panic anymore. It's not okay for the library to explode when there is user-error.

Reply:
Changed to an error.